chainerx_cc/chainerx/index_iterator.h

```diff
@@ -64,10 +64,43 @@ class IndexIterator {
     CHAINERX_HOST_DEVICE void Set(int64_t i) {
         CHAINERX_ASSERT(total_size_ > 0);
         raw_index_ = i;
+#ifdef __CUDA_ARCH__
+        if (total_size_ > 1LL << 31) {
+            // 64-bit division is very slow on GPU
+            size_t a = static_cast<size_t>(i);
+            for (int8_t dim = kNdim; --dim > 0;) {
+                size_t s = static_cast<size_t>(shape_[dim]);
+                if (s & (s - 1)) {
+                    size_t t = a / s;
+                    index_[dim] = a - t * s;
+                    a = t;
+                } else {  // exp of 2
+                    index_[dim] = a & (s - 1);
+                    a >>= __popcll(s - 1);
+                }
+            }
+            index_[0] = a;
+        } else {
+            unsigned int a = static_cast<unsigned int>(i);
```

Comment:
How about `int32_t`?

Reply:
I think `uint32_t` is better because it can store `<2^32` non-negative integer while `int32_t` can store `<2^31`.

Reply:
I am testing it locally now :)